### PR TITLE
WIP: ZMTP heartbeat binding for zeromq/zeromq.js/issues/165

### LINKED ISSUE #165: ZMTP 3.1 heartbeat missing in API
I need to use the [ZMTP 3.1 connection heartbeat](https://rfc.zeromq.org/spec:37/ZMTP/) that is implemented in [zeromq 4.2.0](https://github.com/zeromq/libzmq/releases/tag/v4.2.0).
Since be1ea29a81227ccd0c0f7bd08b1bf77149a3a8c6, libzmq >= 4.2.2 reuns on both linux and windows making this binding is possible, and the API of the nodejs binding actually misses this feature.

### PR DESCRIPTION
**Pull request by vtellier**
This binding should be enough to fix #165 ; but please note that **this hasn't been tested at all before merging!**

### CODE REVIEW

File: binding.cc [hunk 1]
```diff
@@ -1545,6 +1545,11 @@ namespace zmq {
     opts_binary.insert(55); // ZMQ_ZAP_DOMAIN
     opts_int.insert(56); // ZMQ_ROUTER_HANDOVER
     opts_int.insert(66); //ZMQ_HANDSHAKE_IVL
+	#if ZMQ_VERSION_MINOR >= 2
+	opts_int.insert(75); //ZMQ_HEARTBEAT_IVL
+	opts_int.insert(76); //ZMQ_HEARTBEAT_TTL
+	opts_int.insert(77); //ZMQ_HEARTBEAT_TIMEOUT
```
Comment (interpretor) [inline binding.cc]: Shouldn't this block be indented by one more tab?

File: test/socket.js [hunk 2]
```diff
@@ -41,6 +41,13 @@ describe('socket', function(){
   it('should support options', function(){
     sock = zmq.socket('req', { backlog: 30 });
     sock.backlog.should.equal(30);
+	// ZMTP 3.1 heartbeat
+	sock.heartbeat_ivl = 42;
+    sock.heartbeat_ivl.should.equal(42);
+	sock.heartbeat_ttl = 43;
+    sock.heartbeat_ttl.should.equal(43);
+	sock.heartbeat_timeout = 44;
+    sock.heartbeat_timeout.should.equal(44);
```
Comment (interpretor) [inline test/socket.js]: Same wrong indentation here.
Also the wrong test block, as there should be given options in an object:
```js
sock = zmq.socket('req', {
  backlog: 30,
  heartbeat_ivl: 42,
});
```
You could try the `setsockopt()` function and see if the build error still persists. I cannot say atm why the second assignment fails here.
```js
sock.setsockopt(zmq.ZMQ_HEARTBEAT_IVL, 42);
sock.setsockopt(zmq.ZMQ_HEARTBEAT_TTL, 43);
sock.setsockopt(zmq.ZMQ_HEARTBEAT_TIMEOUT, 44);

// Test the options with the getsockopt() function
sock.getsockopt(zmq.ZMQ_HEARTBEAT_TTL).should.equal(43);
// ...

sock.setsockopt('heartbeat_ivl', 42);
sock.setsockopt('heartbeat_ttl', 43);
sock.setsockopt('heartbeat_timeout', 44);

// Same tests here
```

**Comment by vtellier** [comment]
Thanks,
I found it, it's in the documentation of [ZMQ_HEARTBEAT_TTL in libzmq API](http://api.zeromq.org/4-2:zmq-setsockopt).

> any value less than 100 will have no effect.

I fix it and push ASAP

**Comment by interpretor** [comment]
Ok.
I think you don't have to write a test for the socket options, as we don't test every option at the moment.
We could of course test the setting of all available socket options, but I would do this in a separated task.
If it works on your side, it's fine.
Just adjust the indentation and we can merge this.

File: binding.cc (hunk 1 above)
Comment (vtellier) [inline binding.cc]: Yes, I forgot to check the tabs vs spaces before commit.

**Pull request merged**